### ISSUE TITLE
[RFC] Confirm Button decorator 

### ISSUE THREAD

**Issue by fzaninotto** [marmelab/react-admin#2866]
React-admin has an (undocumented') `<Confirm>` component, which lets developer open a confirmation dialog as follows:

```jsx
<Confirm
     isOpen={true}
     title="Delete Item"
     content="Are you sure you want to delete this item?"
     confirm="Yes"
     confirmColor="primary"
     cancel="Cancel"
     onConfirm={() => { // do something }}
     onClose={() => { // do something }}
/>
```

The problem is that developers still have to dispatch actions manually in their own `onConfim` handler. It's quite cumbersome when they just need to make a DeleteButton pessimistic.

I suggest transforming the `Confirm` component so that it accepts a button as child. When provided, this button *replaces* the confirmation button, and dispatches its own action - no need to pass an `onConfirm` handler. This `Confirm` button would handle its own open state, no need to pass it: 

```jsx
<Confirm
     title="Delete Item"
     content="Are you sure you want to delete this item?"
     confirm="Yes"
     confirmColor="primary"
     cancel="Cancel"
>
    <DeleteButton undoable={false} />
</Confirm>
```

When cloning the child, the `Confirm` button could even override the `undoable`, `label`, and `color` props... Also, as the API is a bit different than the (non-documented, but probably used) `Confirm` component, we should probably create another component to keep backwards compatibility.

Any opinion?

**Comment by fzaninotto** [comment]
After playing with the idea, I think it's not a good one.

First, here is my implementation:

```jsx
import React, { Fragment, Component } from 'react';
import Confirm from './Confirm';

class WithConfirm extends Component {
    state = { isOpen: false };
    event = null;
    constructor(props) {
        super(props);
        this.button = React.createRef();
    }

    handleClick = e => {
        this.setState({ isOpen: true });
        e.stopPropagation();
    };

    handleDialogClose = () => {
        this.setState({ isOpen: false });
    };

    handleConfirm = () => {
        this.button.current.firstElementChild.click();
        this.setState({ isOpen: false });
    };

    render() {
        const { children, title, content, ...rest } = this.props;
        return (
            <Fragment>
                <span onMouseDown={this.handleClick}>
                    <span ref={this.button}>
                        {React.cloneElement(children, rest)}
                    </span>
                </span>
                <Confirm
                    isOpen={this.state.isOpen}
                    title={title}
                    content={content}
                    onConfirm={this.handleConfirm}
                    onClose={this.handleDialogClose}
                />
            </Fragment>
        );
    }
}

WithConfirm.defaultProps = {
    title: 'ra.message.are_you_sure',
    content: 'ra.message.irreversible',
};

export default WithConfirm;
```

A large part of the component is just managing the open state of the confirm dialog, which takes a single line in React 9.8 with `useState`. Not a great added value.

What I don't like is that the component must do some black magic to catch the click on the button, then trigger it again on confirm. In fact, this part should not be in a component, it should be a side effect, decorating the click handler of the button. Something like:

```jsx
<Button onClick={withConfirm(this.handleClick)} />
```

Such a side effect is possible with React 9.8, not before. 

Or maybe something like

```jsx
<DeleteButton onClick={confirm} />
```

which is more intuitive when you've been writing code like this in the past:

```html
<a href="/delete" onClick="return alert('are you sure?")>delete</a>
```

The latter syntax would require a change in the `DeleteButton` API, to accept an `onClick` handler. But what handler would developers want to pass that would not be a confirm dialog? 

That leaves us with something like

```jsx
<DeleteButton confirm={<Confirm />} />
```

But my opinion is that this confirm dialog should be implicit when `undoable` is false.

Long story short, I don't think that this component is really worth teaching developers another way to write buttons.

Besides, I personally think that buttons with confirmation should be a material-ui feature, not a react-admin one. But that's another matter.

That leaves us with the current status quo. It's not straightforward to add a confirmation dialog to a button (developers have to [follow the hidden doc](https://marmelab.com/react-admin/List.html#bulk-action-buttons)). So I think the best decision is to ship a `<DeleteWithConfirmationButton>` and a `<BulkDeleteWithConfirmationButton>` component.

**Comment by djhi** [comment]
> React 9.8

? :wink: 

Otherwise :+1: about the conclusion

**Comment by trbhoang** [comment]
<Toolbar {...props}>
        <SaveButton
          label="Approve"
          redirect="edit"
          submitOnEnter={false}
          handleSubmitWithRedirect={
            () => {
              if(!window.confirm('Are you sure?'))
                return false;
              return props.handleSubmitWithRedirect();
            }
          }
        />
      </Toolbar>
